include/xtypes/Module.hpp

```diff
@@ -0,0 +1,404 @@
+/*
+ * Copyright 2019, Proyectos y Sistemas de Mantenimiento SL (eProsima).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef EPROSIMA_XTYPES_MODULE_HPP_
+#define EPROSIMA_XTYPES_MODULE_HPP_
+
+#include <xtypes/xtypes.hpp>
+
+namespace eprosima {
+namespace xtypes {
+
+class Module
+{
+protected:
+    using PairModuleSymbol = std::pair<const Module*, std::string>;
+
+public:
+    Module()
+        : outer(nullptr)
+        , name_("")
+    {}
+
+    Module& create_submodule(
+            const std::string& submodule)
+    {
+        Module* new_module = new Module(this, submodule);
+        auto result = inner.emplace(submodule, new_module);
+        //auto result = inner.emplace(submodule, std::make_shared<Module>(this, submodule));
+        return *result.first->second.get();
+    }
+
+    std::shared_ptr<Module> get_submodule(
+            const std::string& submodule)
+    {
+        return inner[submodule];
+    }
+
+
+    bool has_submodule(
+            const std::string& submodule)
+    {
+        return inner.count(submodule) > 0;
+    }
+
+    Module& operator [] (
+            const std::string& submodule)
+    {
+        return *inner[submodule];
+    }
+
+    bool emplace(
+            std::shared_ptr<Module>&& module)
```

Comment:
If you emplace the shard_ptr is like passing as const &. Maybe you want to use `Args... args` and built the module inside? Should the modules be emplaced, maybe is really small, and if you war passing the reference is like a simple copy.

Reply:
I think this function isn't going to be really used, so I'm going to comment it out with a TODO to remove it completely in the future.